node/actors/bft/src/testonly/run.rs

```diff
@@ -71,77 +110,407 @@ impl Test {
+        // Potentially `ctx::NoCopy` could be used with `port`.
+        let validator_ports = &validator_ports;
+        let sends = &sends;
+        let stores = &stores;
+        let gossip_targets = &gossip_targets;
+        let (gossip_send, gossip_recv) = channel::unbounded();
+
+        // Run networks by receiving from all consensus instances:
+        // * identify the view they are in from the message
+        // * identify the partition they are in based on their network id
+        // * either broadcast to all other instances in the partition, or find out the network
+        //   identity of the target validator and send to it iff they are in the same partition
+        // * simulate the gossiping of finalized blockss
+        scope::run!(ctx, |ctx, s| async move {
+            for (port, recv) in recvs {
+                let gossip_send = gossip_send.clone();
+                s.spawn(async move {
+                    twins_receive_loop(
+                        ctx,
+                        splits,
+                        validator_ports,
+                        sends,
+                        &gossip_targets[&port],
+                        gossip_send,
+                        port,
+                        recv,
+                    )
+                    .await
+                });
             }
-            Network::Mock => {
-                let mut sends = HashMap::new();
-                let mut recvs = vec![];
-                for (i, spec) in specs.iter().enumerate() {
-                    let (actor_pipe, pipe) = pipe::new();
-                    let key = spec.net.validator_key.as_ref().unwrap().public();
-                    sends.insert(key, actor_pipe.send);
-                    recvs.push(actor_pipe.recv);
-                    s.spawn(
-                        async {
-                            let mut pipe = pipe;
-                            spec.run(ctx, &mut pipe).await
+            s.spawn(async { twins_gossip_loop(ctx, stores, gossip_recv).await });
+            anyhow::Ok(())
+        })
+        .await
+    })
+    .await
+}
+
+/// Receive input messages from the consensus actor and send them to the others
+/// according to the partition schedule of the port associated with this instance.
+///
+/// We have to simulate the gossip layer which isn't instantiated by these tests.
+/// If we don't, then if a replica misses a LeaderPrepare message it won't ever get the payload
+/// and won't be able to finalize the block, and won't participate further in the consensus.
+#[allow(clippy::too_many_arguments)]
+async fn twins_receive_loop(
+    ctx: &ctx::Ctx,
+    splits: &PortSplitSchedule,
+    validator_ports: &HashMap<PublicKey, Vec<Port>>,
+    sends: &HashMap<Port, UnboundedSender<io::OutputMessage>>,
+    gossip_targets: &HashSet<Port>,
+    gossip_send: UnboundedSender<(Port, Port, BlockNumber)>,
+    port: Port,
+    mut recv: UnboundedReceiver<io::InputMessage>,
```

Comment:
nit: that's a lot of args, are you sure you don't want to turn it into a method of some struct?